java/rocksjni/portal.h

```diff
@@ -9089,5 +9089,155 @@ class BlockBasedTableOptionsJni
+      return;
+    }
+
+    ROCKSDB_NAMESPACE::WideColumns columns;
+    columns.reserve(columns_len);
+
+    for (int i = 0; i < columns_len; i++) {
+      auto jname = env->GetObjectArrayElement(jNames, i);
+      if (env->ExceptionCheck()) {
+        return;
+      }
+      auto _name = env->GetDirectBufferAddress(jname);
+      if (_name == nullptr) {
+        ROCKSDB_NAMESPACE::RocksDBExceptionJni::ThrowNew(
+            env,
+            "Invalid \"name\" argument (argument is not a valid direct "
+            "ByteBuffer)");
+        return;
+      }
+      auto name = reinterpret_cast<char*>(_name) + namesOffset[i];
+      ROCKSDB_NAMESPACE::Slice nameSlice(name, namesLength[i]);
+
+      auto j_value = env->GetObjectArrayElement(jValues, i);
+      if (env->ExceptionCheck()) {
+        return;
+      }
+      auto _value = env->GetDirectBufferAddress(j_value);
+      if (_value == nullptr) {
+        ROCKSDB_NAMESPACE::RocksDBExceptionJni::ThrowNew(
+            env,
+            "Invalid \"value\" argument (argument is not a valid direct "
+            "ByteBuffer)");
+        return;
+      }
+      auto value = reinterpret_cast<char*>(_value) + valuesOffset[i];
+      ROCKSDB_NAMESPACE::Slice valueSlice(value, valuesLenght[i]);
+
+      columns.push_back(ROCKSDB_NAMESPACE::WideColumn(nameSlice, valueSlice));
+    }
+
+    op(key_slice, columns);
+    return;
+  }
+
+  static void convertWideColumns(
+      std::function<void(ROCKSDB_NAMESPACE::Slice&,
+                         ROCKSDB_NAMESPACE::WideColumns&)>
+          op,
+      JNIEnv* env, jbyteArray jkey, jint keyOffset, jint keyLen,
+      jobjectArray jnames, jobjectArray jvalues) {
+    jbyte* key = new jbyte[keyLen];
+    env->GetByteArrayRegion(jkey, keyOffset, keyLen, key);
+    if (env->ExceptionCheck()) {
+      return;
+    }
+    ROCKSDB_NAMESPACE::Slice key_slice(reinterpret_cast<char*>(key), keyLen);
+
+    const auto column_len = env->GetArrayLength(jnames);
+    rocksdb::WideColumns columns;
+    columns.reserve(column_len);
+
+    std::vector<std::unique_ptr<jbyte[]>> names_vec;
+    names_vec.reserve(column_len);
+
+    std::vector<std::unique_ptr<jbyte[]>> values_vec;
+    values_vec.reserve(column_len);
+
+    for (int i = 0; i < column_len; i++) {
+      auto jname =
+          static_cast<jbyteArray>(env->GetObjectArrayElement(jnames, i));
+      if (env->ExceptionCheck()) {
+        return;
+      }
+      auto jname_len = env->GetArrayLength(jname);
+      auto name = std::make_unique<jbyte[]>(jname_len);
+      env->GetByteArrayRegion(jname, 0, jname_len, name.get());
+      if (env->ExceptionCheck()) {
+        return;
+      }
+      ROCKSDB_NAMESPACE::Slice name_slice(reinterpret_cast<char*>(name.get()),
+                                          jname_len);
+      names_vec.push_back(std::move(name));
+
+      auto j_value =
+          static_cast<jbyteArray>(env->GetObjectArrayElement(jvalues, i));
+      if (env->ExceptionCheck()) {
+        return;
```

Comment:
Do we need to free the JNI local references for name/jname here?

Reply:
No we don't need to. [doc](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#GetObjectArrayElement)

Reply:
As I understand it, `GetObjectArrayElement` returns a `jobject` and when you are finished with this (on either success or error paths), you should call `DeleteLocalRef` on that object. There are other examples of this pattern already in the existing code-base, for example: https://github.com/facebook/rocksdb/blob/main/java/rocksjni/rocksjni.cc#L462